src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,374 @@
+                    dest.SetByte(j + 1, alphabet[(src.GetByte(i) & 0x03) << 4]);
+                    dest.SetByte(j + 2, EQUALS_SIGN);
+                    dest.SetByte(j + 3, EQUALS_SIGN);
+                    j += 4;
+                    break;
+            }
+            //restore src's ReaderIndex and WriterIndex
+            src.SetIndex(srcReaderIndex, srcWriterIndex);
+            //remove last byte if it's NewLine
+            int destLength = dest.GetByte(j - 1) == NEW_LINE ? j - 1 : j;
+            return destLength;
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator allocator)
+        {
+            if (src == null)
+            {
+                throw new ArgumentNullException(nameof(src));
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException(nameof(dialect.alphabet));
+            }
+            Contract.Assert(dialect.alphabet.Length == 64, "alphabet.Length must be 64!");
+            if (offset < src.ReaderIndex || offset + length > src.ReaderIndex + src.ReadableBytes)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+            if (length <= 0)
+            {
+                return Unpooled.Empty;
+            }
+
+            int outLength = length / 3 * 4;
+            outLength += length % 3 > 0 ? 4 : 0;
+            outLength += breakLines ? outLength / MAX_LINE_LENGTH : 0;
+            IByteBuffer dest = allocator.Buffer(outLength).WithOrder(src.Order);
+            int destLength = 0;
+            fixed (byte* alphabet = dialect.alphabet)
+            {
+                if (src.IoBufferCount == 1 && dest.IoBufferCount == 1)
+                {
+                    destLength = EncodeUsingPointer(alphabet, src, dest, offset, length, breakLines);
+                }
+                else
+                {
+                    destLength = EncodeUsingGetSet(alphabet, src, dest, offset, length, breakLines);
+                }
+            }
+            return dest.SetIndex(0, destLength);
+        }
+
+
+        public static IByteBuffer Decode(IByteBuffer src)
+        {
+            return Decode(src, Base64Dialect.STANDARD);
+        }
+
+        public static IByteBuffer Decode(IByteBuffer src, Base64Dialect dialect)
+        {
+            return Decode(src, src.ReaderIndex, src.ReadableBytes, dialect);
+        }
+
+        public static IByteBuffer Decode(IByteBuffer src, int offset, int length, Base64Dialect dialect)
+        {
+            return Decode(src, offset, length, dialect, src.Allocator);
+        }
+
+        private unsafe static int DecodeUsingPointer(IByteBuffer src, IByteBuffer dest, sbyte* decodabet, int offset, int length)
+        {
+            int charCount = 0;
+            Contract.Assert(dest.WriterIndex == 0 && dest.ArrayOffset == 0);
```

Comment:
>dest.ArrayOffset == 0 [](start = 53, length = 21)

This may not be true. E.g. for PooledByteBuffer retrieved from PooledByteBufferAllocator (most common high perf buffer implementation) ArrayOffset will most likely be non-zero.